src/Tempest/Http/src/Route.php

```diff
@@ -28,12 +29,12 @@ public function __construct(
          */
         public array $middleware = [],
     ) {
-        // Routes can have parameters in the form of "/{PARAM}/",
-        // these parameters are replaced with a regex matching group
-        $matchingRegex = preg_replace(
-            '#\{(\w+)}#',
-            '([^/]++)',
-            $uri
+
+        // Routes can have parameters in the form of "/{PARAM}/" or /{PARAM:CUSTOM_REGEX},
+        // these parameters are replaced with a regex matching group or with the custom regex
+        $matchingRegex = (string)str($this->uri)->replaceRegex(
+            '#\{(?<name>\w+)(?::(?<regex>[^}]+))?\}#',
```

Comment:
@vsergiu93 - I think there is a bug here. If you can fix that, I will merge this and we can address the other pieces in future PRs.

Reply:
@aidan-casey I'll have a look thanks

Reply:
> What happens with something someone shouldn't do like:
> 
> ```regexp
> /user/{name: {name}}
> ```
 
Well, you'll get some malformed $matchingRegex, `/user/( {name)}` this is obviously incorrect, and I think here we need to see how to approach this. From DX pov I would say it would be nice to throw some exceptions, but that means we need to validate the custom regex therefore performance might be impacted (we could maybe conditionally do this check only in development and disable this in prod).

I guess we could check how other frameworks handle this part, see what kind of checks they do if any, if there are any restrictions in place.

Also @aidan-casey, I think I would like to write some unit tests for src/Tempest/Http/src/Route.php, I mean with the integrations tests is quite cumbersome to test this part with custom regex properly. Where should I put this tests?

Reply:
> and I think here we need to see how to approach this

My inclination would be to just ignore nested brackets. See [here](https://github.com/nikic/FastRoute/blob/master/src/RouteParser/Std.php) for inspiration.

> I think I would like to write some unit tests for src/Tempest/Http/src/Route.php

Brilliant idea! I would place these in `src/Tempest/Http/tests/RouteTest.php`.

Reply:
@aidan-casey Thanks for the link, yeah, as I was suspecting already, ideally some checks need to be put in place, like 
not allowing users to have something like this `/{lang:(en|de)}` or this `/foo/{test}/{test:\d+}`. 
I guess it would make sense to add some checks and throw some exceptions to improve DX, but not sure if we should do it in this PR.

I did some improvements (basically I reused some of the regex from the packaged you linked), you could check the unit test input and expected for different formats and let me know if is ok. 

One more thing I noticed that FastRoute allows for route segment name to be surrounded  by spaces `/foo/{ param : \d{1,9} }` , I did add a trim for the custom regex, do you think we should handle this case as well?

Reply:
Let's get this merged and revisit in some follow-up PRs. I think these deserve some broader discussion.